processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java

```diff
@@ -257,4 +264,52 @@ public Object apply(final Result<SelectResultValue> input)
   {
     return intervalChunkingQueryRunnerDecorator.decorate(runner, this);
   }
+
+  @Override
+  public <T extends LogicalSegment> List<T> filterSegments(SelectQuery query, List<T> segments)
+  {
+    PagingSpec pagingSpec = query.getPagingSpec();
+    Map<String, Integer> paging = pagingSpec.getPagingIdentifiers();
+    if (paging == null || paging.isEmpty()) {
+      return segments;
+    }
+    List<Interval> intervals = Lists.newArrayList(
+        Iterables.transform(
+            paging.keySet(),
+            SegmentDesc.INTERVAL_EXTRACTOR
+        )
+    );
+    Collections.sort(
+        intervals, new Comparator<Interval>()
+        {
+          @Override
+          public int compare(Interval o1, Interval o2)
+          {
+            return Longs.compare(o1.getStartMillis(), o2.getStartMillis());
+          }
+        }
+    );
+
+    List<T> queryIntervals = Lists.newArrayList(segments);
+
+    Iterator<T> it = queryIntervals.iterator();
+    if (query.isDescending()) {
+      final long lastEnd = intervals.get(intervals.size() - 1).getEndMillis();
+      while (it.hasNext()) {
+        T segment = it.next();
+        if (segment.getInterval().getStartMillis() > lastEnd) {
+          it.remove();
+        }
+      }
+    } else {
+      final long firstStart = intervals.get(0).getStartMillis();
+      while (it.hasNext()) {
+        T segment = it.next();
+        if (segment.getInterval().getEndMillis() < firstStart) {
+          it.remove();
+        }
+      }
+    }
```

Comment:
does this handles the case when segments overlaps in intervals ? 
e.g 
- Segment 1 - 2016-01-01T00:00:00Z/2016-01-02T00:00:00Z - version v1 
- Segment 2 - 2016-01-01T01:00:00Z/2016-01-01T02:00:00Z  - version v2 > v1

I remember seeing issues with similar case in timeboundary query.